### ISSUE TITLE
20.03 -> 20.09: Unexpected change in behaviour of wpa_supplicant

### ISSUE THREAD

**Issue by fadenb** [NixOS/nixpkgs#102457]
**Describe the bug**
I was using a manually created wpa_supplicant config (placed at `/etc/wpa_supplicant.conf`) containing all my wireless networks on NixOS 20.03 combined with the following block in my `configuration.nix`:
```nix
    wireless = {
      enable = true;
      interfaces = [ "wlp3s0"];
      userControlled.enable = true;
      userControlled.group = "wheel";
    };
```
NixOS started the wpa supplicant and everything worked as expected.
After updating to 20.09 NixOS is generating its own config file for wpa_supplicant which is mostly empty (only contains `ctrl_interface` and `update_config`) resulting in no wifi connectivity.

**To Reproduce**
* Be on NixOS 20.03
* Use config snippet from section above
* Add `nixos` channel pointing to the URL for 20.09 channel
* do a `nixos-rebuild boot --upgrade`
* reboot
* Notice no more wifi

**Expected behavior**
I expected that the wpa supplicant uses the existing `/etc/wpa_supplicant.conf` as it did on 20.03 or a warning that the behaviour changed.

**Notify maintainers**

@bobismijnnaam ( https://github.com/NixOS/nixpkgs/commit/71ea6a9a41f1340e6821812f8b7e331072dbd020 )

**Metadata**

 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.72, NixOS, 20.09.1500.edb26126d98 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(root): `"nixos-20.09.1500.edb26126d98, nixos-unstable-small-21.03pre248327.278d8641c8d, nixosunstable-21.03pre248062.5265d49a36b"`
 - channels(<MY-USERNAME>): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module: networking.wireless
```

**Comment by rnhmjoj** [comment]
To be fair, it seems (8e3da733b15a3c835305d0ba99e33ab33b4ce2d1) it is already warning that the file is not being used.
In any case, I don't it's a good idea to change /etc manually while using NixOS. If you are doing so because of the concern of storing the passwords/PSK inside the Nix store, I guess it's a good time to add an option for external storage of password.

**Comment by fadenb** [comment]
> In any case, I don't it's a good idea to change /etc manually while using NixOS.

I fully agree but reality sometimes is different from best practices 😉 

> To be fair, it seems (8e3da73) there is already warning that the file is not being used.

The warning is being generated when it is already too late but it surely helps to figure out what is wrong. Perhaps we can extend it a bit to explain what conditions trigger the use of a generated config?


My issue is more with that the behaviour changed unexpectely.
I did not find any reference to the change in the release notes. Should I propose some addendum as a PR to warn other users?


> If you are doing so because of the concern of storing the password/PSKs inside the Nix store, I guess it's a good time to add an option for external storage of passwords.

Not at all as my primary usecase with wifi is a single-user laptop :)
I often find myself in the situation to need ad-hoc changes to the config to get some connectivity. Is there perhaps a feasible way to do that while retaining the generated config? For example like having the generated config dynamically include my other file?

**Comment by rnhmjoj** [comment]
> Perhaps we can extend it a bit to explain what conditions trigger the use of a generated config?

The problem is that enabling a module and manually changing the configuration shouldn't be expected to work, at all.
I wouldn't mention the conditions because they may change again unexpectedly (or change to "always"); also doing so
would suggest something against the core idea of NixOS.

> I did not find any reference to the change in the release notes. Should I propose some addendum as a PR to warn other users?

Even if unintentionally, we broke compatibility, so a word of warning is necessary in my opinion. Please go on.

> Is there perhaps a feasible way to do that while retaining the generated config? For example like having the generated config dynamically include my other file?

I don't think so but it's certainly possible to add some code to the `preStart` to concatenate a file with the generated one before starting the service.

> I often find myself in the situation to need ad-hoc changes to the config to get some connectivity.

My solution is use wpa_cli: it can do anything the configuration file can but the settings are lost if you restart the daemon.
If I need to apply the same changes more than once, I put the commands into a text file and pipe it into wpa_cli:
```
echo '
add_network 8
set_network 8 ssid "..."
set_network 8 pask "..."
set_network 8 key_mgnt WPA-PSK
enable_network 8 
' > settings
wpa_cli -i wlan0 < settings
```

**Comment by bobismijnnaam** [comment]
Just chipping in my two cents of understanding on why the changed behavior is useful: if your expected behavior were the case, `wpa_cli` would not work as the `ctrl_interface` information is not present in your hand-crafted `wpa_supplicant.conf`. So setting `userControlled` to `true` really requires a nix-created `wpa_supplicant.conf`. If you never use `wpa_cli` however you don't notice it (which is how I ran into this problem in the first place).

In any case, the important part is that this change in behavior should technically be documented, if I understand correctly. I am not familiar enough with nixos practices regarding this. Is the change significant enough to be included in the "backwards incompatible changes" section of the release notes? Who do we ask to determine this? And if so, can we include changes to the release notes even after 20.09 was released? Another question, are there any guidelines w.r.t. breaking behavior in the contributing how to guide? Because if so, I missed it. If not, maybe a section should be added on breaking behavior...? Do we discuss this on the forum or should we make an issue on nixpkgs?

As a final more on-topic idea, it _is_ bit fishy that the only way to figure out if a `wpa_supplicant.conf` is being generated is to read the nix code. To improve this, we could add a "prevent generating config file" flag, which ensure your nixos installation fails if you have a) enabled the "prevent generating config file" flag and b) enabled any options that would normally cause a config file to be generated. That way we give users who do not want a config file an option to be alerted when they're about to shadow their own config file. People who don't use the config file can ignore that option and turn things on and off as they please. Not sure if it is very nix-like, but it would prevent vague issues like these in the future.

**Comment by socksy** [comment]
If anyone, like me, was looking for the way to get back the old behaviour (until it's possible to have well supported secrets in the nix conf I'd rather not manage my `/etc/wpa_supplicant.conf` — I store those on tarsnap), then it seems that removing `userControlled.enable = true;` is the way forward.

I would strongly support suggestions to document breaking changes in the release notes :+1:

**Comment by fadenb** [comment]
@socksy still on my todo list. Hopefully later today I'll propose a pull request.

**Comment by rnhmjoj** [comment]
I merged PR #104589 and backported it to 20.09. I think this can be closed.
